stability: prepare protos for proposer-evaluted KV

tbg opened the pull request:
As a page-in task for proposer-evaluted KV, this establishes structures
to be used for sending the new types of proposals over the wire. In short,
they consist of three components:
- Local state only available to the leader (intents to resolve, ...)
- A replicated state update for all Replicas (split/merge triggers,
  range descriptor updates, log queue truncation parameters, increased
  GC thresholds, ...). This always contains the new applied index and
  stats, so the structure is intentionally not optional (but most of
  its contents are, so it is not a large struct and serializes well enough).
- A RocksDB WriteBatch (i.e. the actual data to be written to disk).

There are some TODOs I would like to address before merging (moving code,
adding tests and assertions), but this should make for a good high-level
review as is.

There are a lot of changes; I recommend focusing on
`storage/replica_proposal.go` as well as `storage/storagebase`.

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/9883)

<!-- Reviewable:end -->

tschottdorf commented on storage/replica_proposal.go:
```diff
@@ -431,37 +490,51 @@ func (r *Replica) handleTrigger(
 
-	if trigger.computeChecksum != nil {
-		r.computeChecksumTrigger(ctx, *trigger.computeChecksum)
+	if !reflect.DeepEqual(pd, ProposalData{}) {
```
I'll clean this up. I'll also make sure that we don't always call `r.assertState` before this merges (which is done elsewhere).

tschottdorf commented:
Rebased on top of node liveness (#9530).

petermattis commented:
Generally, LGTM, though I wish this was showing me diffs between `replica_trigger.go` and `replica_proposal.go`. Perhaps rename `replica_proposal.go` temporarily to `replica_trigger.go`? Without that, it is hard to see if there are substantive differences between `replica_trigger.go` and `replica_proposal.go`. 

---

Review status: 0 of 19 files reviewed at latest revision, 5 unresolved discussions, all commit checks successful.

---

*[storage/replica_proposal.go, line 537 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9883#-KTogjyrENJQvunbTC3G-r1-537:-KTtD5cWH6d2GkGOTfOt:bupr9kb) ([raw file](https://github.com/cockroachdb/cockroach/blob/028b1d0071ca6aa5613dc831797b68b2aebd1066/storage/replica_proposal.go#L537)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> I'll clean this up. I'll also make sure that we don't always call `r.assertState` before this merges (which is done elsewhere).
</details>
Do you need `DeepEqual` here? Seems sufficient to do `pd != ProposalData{}`.

---

*[storage/replica_proposal.go, line 74 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/9883#-KTtBoTiJUbur8qJvX64:-KTtBoTji85VJhncVflY:bea1e99) ([raw file](https://github.com/cockroachdb/cockroach/blob/71bd7c09685cffe1429d6b23990b76ed569c5498/storage/replica_proposal.go#L74)):*
> ```Go
> // ProposalData is the result of preparing a Raft proposal. That is, the
> // proposer (which holds the lease, at least in the case in which the command
> // will complete successfully) has simulated execution of the result and is
> ```

`simulated` feels wrong here. The proposer has executed but not committed a command.

---

*[storage/replica_proposal.go, line 78 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/9883#-KTtC-Incd9-z0vm79Zj:-KTtC-Incd9-z0vm79Zk:b-xwnmq2) ([raw file](https://github.com/cockroachdb/cockroach/blob/71bd7c09685cffe1429d6b23990b76ed569c5498/storage/replica_proposal.go#L78)):*
> ```Go
> //
> // a) changes to be written to disk when applying the command
> // b) changes to the state which may require special handling (i.e. code
> ```

All disk state should be contained in part a), correct? `s/state/in-memory state/g`

---

*[storage/replica_proposal.go, line 105 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/9883#-KTtCNhAnQztCAXYhKhW:-KTtCNhAnQztCAXYhKhX:b-rvmltu) ([raw file](https://github.com/cockroachdb/cockroach/blob/71bd7c09685cffe1429d6b23990b76ed569c5498/storage/replica_proposal.go#L105)):*
> ```Go
> // TODO(tschottdorf): write tests before merge to quickly catch forgetting
> // to add new fields.
> func (p *ProposalData) Absorb(q ProposalData) error {
> ```

I'm not too fond of the term `Absorb` here. Merge? Update?

---

*[storage/replica_proposal.go, line 189 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/9883#-KTtClqxX-0QVP52pH09:-KTtClqxX-0QVP52pH0A:b-okqn0y) ([raw file](https://github.com/cockroachdb/cockroach/blob/71bd7c09685cffe1429d6b23990b76ed569c5498/storage/replica_proposal.go#L189)):*
> ```Go
> 	coalesce := func(lhs *bool, rhs bool) {
> 		*lhs = *lhs || rhs
> 	}
> ```

I've always wondered if there is overhead to using a closure like this vs making this a package-level function. Do you know?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9883#-:-KTtDRa2OG1bkM4-rQZp:b-a81ub8)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Review status: 0 of 19 files reviewed at latest revision, 5 unresolved discussions, all commit checks successful.

---

*[storage/replica_proposal.go, line 189 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/9883#-KTtClqxX-0QVP52pH09:-KTtXaFgJCP08gep0SSm:buunbnk) ([raw file](https://github.com/cockroachdb/cockroach/blob/71bd7c09685cffe1429d6b23990b76ed569c5498/storage/replica_proposal.go#L189)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> I've always wondered if there is overhead to using a closure like this vs making this a package-level function. Do you know?
</details>
I know now:

```
BenchmarkClosureAlloc-4        	500000000      	         3.11 ns/op    	       0 B/op  	       0 allocs/op
BenchmarkClosureFunction-4     	2000000000     	         1.44 ns/op    	       0 B/op  	       0 allocs/op
BenchmarkClosureInline-4       	2000000000     	         1.39 ns/op    	       0 B/op  	       0 allocs/op
PASS
```

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9883)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Review status: 0 of 19 files reviewed at latest revision, 5 unresolved discussions, all commit checks successful.

---

*[storage/replica_proposal.go, line 189 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/9883#-KTtClqxX-0QVP52pH09:-KTtXm1UaZqKJgE-5qto:b81o1w0) ([raw file](https://github.com/cockroachdb/cockroach/blob/71bd7c09685cffe1429d6b23990b76ed569c5498/storage/replica_proposal.go#L189)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> I know now:
> 
> ```
> BenchmarkClosureAlloc-4        	500000000      	         3.11 ns/op    	       0 B/op  	       0 allocs/op
> BenchmarkClosureFunction-4     	2000000000     	         1.44 ns/op    	       0 B/op  	       0 allocs/op
> BenchmarkClosureInline-4       	2000000000     	         1.39 ns/op    	       0 B/op  	       0 allocs/op
> PASS
> ```
</details>
(github.com/tschottdorf/goplay/closure_alloc)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9883)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Oops, sorry about that. Pushed a (temporary) update.

---

Review status: 0 of 19 files reviewed at latest revision, 5 unresolved discussions, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9883#-:-KTtY2RKzycR6Bd9IuH1:bacz06i)*
<!-- Sent from Reviewable.io -->

petermattis commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 0 of 18 files reviewed at latest revision, 12 unresolved discussions, all commit checks successful.

---

*[storage/replica.go, line 2668 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/9883#-KTtwntbPPRm2lrIkZOy:-KTtwntbPPRm2lrIkZOz:b-2i62bo) ([raw file](https://github.com/cockroachdb/cockroach/blob/2ac09ca5612f018f081db4f4f2e4ed238f19d92c/storage/replica.go#L2668)):*
> ```Go
> 	// Handle the ProposalData, executing any side effects of the last
> 	// state machine transition.
> 	r.handleProposalData(ctx, raftCmd.OriginReplica, pd)
> ```

Looks like you're calling this even if the batch commit fails. I think that's fine, but perhaps call this out in comments. 

---

*[storage/replica_command.go, line 1779 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/9883#-KTtx9gnbLwTMMkjWcf0:-KTtx9gnbLwTMMkjWcf1:big597a) ([raw file](https://github.com/cockroachdb/cockroach/blob/2ac09ca5612f018f081db4f4f2e4ed238f19d92c/storage/replica_command.go#L1779)):*
> ```Go
> 	t := true
> 	pd.leaseMetricsResult = &t
> 	pd.maybeGossipNodeLiveness = &keys.NodeLivenessSpan
> ```

Why is `maybeGossipNodeLiveness` a span and not a boolean like the other `maybeGossip` fields?

---

*[storage/replica_command.go, line 2001 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/9883#-KTtxJ6ulXa8SYPvpMRS:-KTtxJ6ulXa8SYPvpMRT:b-az46mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/2ac09ca5612f018f081db4f4f2e4ed238f19d92c/storage/replica_command.go#L2001)):*
> ```Go
> 	}
> 	var pd ProposalData
> 	pd.ComputeChecksum = &args
> ```

`pd := ProposalData{ComputeChecksum: &args}`

---

*[storage/replica_command.go, line 2778 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/9883#-KTtxTHzY-UcijsVNbt_:-KTtxTHzY-UcijsVNbta:bavbllr) ([raw file](https://github.com/cockroachdb/cockroach/blob/2ac09ca5612f018f081db4f4f2e4ed238f19d92c/storage/replica_command.go#L2778)):*
> ```Go
> 	rightDeltaMS := bothDeltaMS
> 	rightDeltaMS.Subtract(leftDeltaMS)
> 	var pd ProposalData
> ```

```
pd := ProposalData{
  BlockReads: true,
  Split: &storabebase.Split{...},
}
```

There are a handful of other places you could be using composite literal syntax like this. Any reason you're not?

---

*[storage/replica_trigger.go, line 189 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/9883#-KTtClqxX-0QVP52pH09:-KTturuS4E0AOoGYe_cX:b-ucfkn4) ([raw file](https://github.com/cockroachdb/cockroach/blob/71bd7c09685cffe1429d6b23990b76ed569c5498/storage/replica_trigger.go#L189)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> (github.com/tschottdorf/goplay/closure_alloc)
</details>
Thanks.

---

*[storage/replica_trigger.go, line 420 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/9883#-KTtxqFA4EDRc7t6Qjux:-KTtxqFA4EDRc7t6Qjuy:b-gaacpe) ([raw file](https://github.com/cockroachdb/cockroach/blob/2ac09ca5612f018f081db4f4f2e4ed238f19d92c/storage/replica_trigger.go#L420)):*
> ```Go
> 	pd.State.Frozen = storagebase.ReplicaState_FrozenEnum(0)
> 
> 	// TODO(tschottdorf): check this without reflect.
> ```

What is this comment about? What `reflect`?

---

*[storage/replica_trigger.go, line 456 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/9883#-KTtxuuvEmj4PbUbNRCV:-KTtxuuwlT6V-OJMVROz:b-we1n26) ([raw file](https://github.com/cockroachdb/cockroach/blob/2ac09ca5612f018f081db4f4f2e4ed238f19d92c/storage/replica_trigger.go#L456)):*
> ```Go
> 		r.store.raftEntryCache.clearTo(
> 			r.RangeID, newTruncState.Index+1,
> 		)
> ```

This could be 1 line vs 3.

---

*[storage/store.go, line 1572 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/9883#-KTty5n7sgn1ELueKevr:-KTty5n7sgn1ELueKevs:bezv7xl) ([raw file](https://github.com/cockroachdb/cockroach/blob/2ac09ca5612f018f081db4f4f2e4ed238f19d92c/storage/store.go#L1572)):*
> ```Go
> 	// the EndKey of the LHS replica and also adds the RHS replica
> 	// to the store's replica map.
> 	log.Warningf(context.TODO(), "rightRng = %s", rightRng)
> ```

I assume you're going to remove these log messages before merging.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9883#-:-KTtyBcAk34veIawxV4_:bnfp4nl)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Review status: 0 of 19 files reviewed at latest revision, 12 unresolved discussions, some commit checks pending.

---

*[storage/replica.go, line 2668 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/9883#-KTtwntbPPRm2lrIkZOy:-KTxVKKoB8aqBczR-EW9:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/2ac09ca5612f018f081db4f4f2e4ed238f19d92c/storage/replica.go#L2668)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> Looks like you're calling this even if the batch commit fails. I think that's fine, but perhaps call this out in comments. 
</details>
Done.

---

*[storage/replica_command.go, line 2001 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/9883#-KTtxJ6ulXa8SYPvpMRS:-KTxQ4p3GufvsZLlrq38:b-pnhr4h) ([raw file](https://github.com/cockroachdb/cockroach/blob/2ac09ca5612f018f081db4f4f2e4ed238f19d92c/storage/replica_command.go#L2001)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> `pd := ProposalData{ComputeChecksum: &args}`
</details>
See below.

---

*[storage/replica_command.go, line 2778 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/9883#-KTtxTHzY-UcijsVNbt_:-KTxPrqBooZfKw7MEyX2:b56c0su) ([raw file](https://github.com/cockroachdb/cockroach/blob/2ac09ca5612f018f081db4f4f2e4ed238f19d92c/storage/replica_command.go#L2778)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> ```
> pd := ProposalData{
>   BlockReads: true,
>   Split: &storabebase.Split{...},
> }
> ```
> 
> There are a handful of other places you could be using composite literal syntax like this. Any reason you're not?
</details>
It would have to be

```go
	pd := ProposalData{
	  ReplicatedProposalData: ReplicatedProposalData{
		BlockReads: true,
	  }
	}

```

which is really noisy.

---

*[storage/store.go, line 1572 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/9883#-KTty5n7sgn1ELueKevr:-KTxUjqj4DYbtnMwuSNz:b-tk3w4) ([raw file](https://github.com/cockroachdb/cockroach/blob/2ac09ca5612f018f081db4f4f2e4ed238f19d92c/storage/store.go#L1572)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> I assume you're going to remove these log messages before merging.
</details>
Now that you've reminded me that they're there, I might.
Done.

---

*[storage/replica_trigger.go, line 537 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9883#-KTogjyrENJQvunbTC3G-r1-537:-KTtaVXFGaFdJf-VYUzI:be9v2ye) ([raw file](https://github.com/cockroachdb/cockroach/blob/028b1d0071ca6aa5613dc831797b68b2aebd1066/storage/replica_trigger.go#L537)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> Do you need `DeepEqual` here? Seems sufficient to do `pd != ProposalData{}`.
</details>
The structs couldn't be compared directly because of the nested `[]intentsWithArg`, but I've made that `*[]intentsWithArg` which is mildly awkward but allows the comparison, which I think is worth the boilerplate it avoids.

---

*[storage/replica_trigger.go, line 74 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/9883#-KTtBoTiJUbur8qJvX64:-KTtZviEsyCMvCG68LZ7:b-rk3ic7) ([raw file](https://github.com/cockroachdb/cockroach/blob/71bd7c09685cffe1429d6b23990b76ed569c5498/storage/replica_trigger.go#L74)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> `simulated` feels wrong here. The proposer has executed but not committed a command.
</details>
I explicitly avoided `executed` because that  will give you the wrong impression. How about "evaluated"?

---

*[storage/replica_trigger.go, line 78 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/9883#-KTtC-Incd9-z0vm79Zj:-KTt_3xQ3CxrV8LeZMtU:b2wlf1p) ([raw file](https://github.com/cockroachdb/cockroach/blob/71bd7c09685cffe1429d6b23990b76ed569c5498/storage/replica_trigger.go#L78)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> All disk state should be contained in part a), correct? `s/state/in-memory state/g`
</details>
No, also disk state. For example, when we get a new descriptor (as part of a replica change, for example) we need to do a lot more than update our in-memory state. I decided that it was too fragile and costly to inspect each and every write set for changes to disk state that would have to trigger certain actions. Instead, I set it up so that we can (mostly) attach any part of a `ReplicaState` to the batch when it changes (along with auxiliary data required, such as the triggers), making that mechanism explicit. I think that's the right way to go, but I'd like to hear your opinion.

---

*[storage/replica_trigger.go, line 105 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/9883#-KTtCNhAnQztCAXYhKhW:-KTt_itexk_Deq6YQS-_:b-8ga0nj) ([raw file](https://github.com/cockroachdb/cockroach/blob/71bd7c09685cffe1429d6b23990b76ed569c5498/storage/replica_trigger.go#L105)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> I'm not too fond of the term `Absorb` here. Merge? Update?
</details>
I want to make absolutely clear that the RHS is not supposed to be used any more, which I don't think the others do quite as clearly. What do you not like about `Absorb`?

---

*[storage/replica_trigger.go, line 189 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/9883#-KTtClqxX-0QVP52pH09:-KTxTZYB-lEZ37PdrdSs:bkrxkie) ([raw file](https://github.com/cockroachdb/cockroach/blob/71bd7c09685cffe1429d6b23990b76ed569c5498/storage/replica_trigger.go#L189)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> Thanks.
</details>
Actually I lied. This is what it should look like:

```
BenchmarkClosureAlloc-4        	20000000       	       103 ns/op       	       1 B/op  	       1 allocs/op
BenchmarkClosureFunction-4     	2000000000     	         1.07 ns/op    	       0 B/op  	       0 allocs/op
BenchmarkClosureInline-4       	2000000000     	         0.30 ns/op    	       0 B/op  	       0 allocs/op
PASS
ok     	github.com/tschottdorf/goplay/closure_alloc    	5.095s
```

My previous code declared `lhs` outside of the `b.N` loop, hiding the fact that it goes on the heap.

Moved this out to a static function.

---

*[storage/replica_trigger.go, line 420 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/9883#-KTtxqFA4EDRc7t6Qjux:-KTxU_RdQ76c8mV9OQMc:b-fvf91t) ([raw file](https://github.com/cockroachdb/cockroach/blob/2ac09ca5612f018f081db4f4f2e4ed238f19d92c/storage/replica_trigger.go#L420)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> What is this comment about? What `reflect`?
</details>
Detritus, removed.

---

*[storage/replica_trigger.go, line 456 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/9883#-KTtxuuvEmj4PbUbNRCV:-KTxUee8mW-1HB2SxFt3:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/2ac09ca5612f018f081db4f4f2e4ed238f19d92c/storage/replica_trigger.go#L456)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> This could be 1 line vs 3.
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9883)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Review status: 0 of 19 files reviewed at latest revision, 12 unresolved discussions, some commit checks pending.

---

*[storage/replica_command.go, line 1779 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/9883#-KTtx9gnbLwTMMkjWcf0:-KTxWx2dM-5wXEyiopNZ:b6icdxr) ([raw file](https://github.com/cockroachdb/cockroach/blob/2ac09ca5612f018f081db4f4f2e4ed238f19d92c/storage/replica_command.go#L1779)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> Why is `maybeGossipNodeLiveness` a span and not a boolean like the other `maybeGossip` fields?
</details>
`maybeGossipNodeLiveness` takes a span argument (roughly, it only gossips if it's the leader, but will also only gossip for the key range which either belongs to a single node's record or all records).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9883)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
TFTRs. I'm sure there are more comments to be made, which I'm happy to address in a follow-up.

---

Review status: 0 of 19 files reviewed at latest revision, 18 unresolved discussions.

---

*[pkg/storage/replica.go, line 2841 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/9883#-KTxXYhGgXwJb_UknEaq:-KTxXYhGgXwJb_UknEar:b-1tr683) ([raw file](https://github.com/cockroachdb/cockroach/blob/2da5711d902fa5ad6179a198c80368074934bc5d/pkg/storage/replica.go#L2841)):*
> ```Go
> 	// Flush the MVCC stats to the batch. Note that we need to grab the previous
> 	// stats now, for they might've been changed due to side effects.
> 	// TODO(tschottdorf): refactor that for clarity. Also doesn't seem true
> ```

Fixed up the assertion here.

---

*[pkg/storage/replica.go, line 2850 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/9883#-KTxXXaxzkM26QJeThCA:-KTxXXaxzkM26QJeThCB:bqd1ksg) ([raw file](https://github.com/cockroachdb/cockroach/blob/2da5711d902fa5ad6179a198c80368074934bc5d/pkg/storage/replica.go#L2850)):*
> ```Go
> 	pd.State.Stats = newMS
> 	// TODO(tschottdorf): return delta up the stack as separate variable.
> 	// It's used by the caller.
> ```

I realized that this is actually best kept in LocalProposalData. Moved it there.

---

*[pkg/storage/replica.go, line 2933 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/9883#-KTxX_f35HRwiP-9A5bK:-KTxX_f35HRwiP-9A5bL:b-j59qev) ([raw file](https://github.com/cockroachdb/cockroach/blob/2da5711d902fa5ad6179a198c80368074934bc5d/pkg/storage/replica.go#L2933)):*
> ```Go
> 	}
> 
> 	// TODO(tschottdorf): assert that this has zero delta at this point
> ```

Added this assertion.

---

*[pkg/storage/replica_proposal.go, line 69 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/9883#-KTxXwV8X77yhHqewdpw:-KTxXwV8X77yhHqewdpx:bt2vr1b) ([raw file](https://github.com/cockroachdb/cockroach/blob/2da5711d902fa5ad6179a198c80368074934bc5d/pkg/storage/replica_proposal.go#L69)):*
> ```Go
> 	// Call maybeAddToReplicaGCQueue.
> 	addToReplicaGCQueue bool
> 	// Call TODO
> ```

Fixed.

---

*[pkg/storage/replica_state.go, line 91 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/9883#-KTxYGCAyTtYCeb_4XrI:-KTxYGCAyTtYCeb_4XrJ:b-225q4t) ([raw file](https://github.com/cockroachdb/cockroach/blob/2da5711d902fa5ad6179a198c80368074934bc5d/pkg/storage/replica_state.go#L91)):*
> ```Go
> // TODO(tschottdorf): test and assert that none of the optional values are
> // missing when- ever saveState is called. Optional values should be reserved
> // strictly for use in ProposalData. Do before merge.
> ```

I've done this but it's worth an extra PR.

---

*[pkg/storage/replica_state.go, line 299 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/9883#-KTxYIIp4JM4saEfSZoV:-KTxYIIp4JM4saEfSZoW:b34uopr) ([raw file](https://github.com/cockroachdb/cockroach/blob/2da5711d902fa5ad6179a198c80368074934bc5d/pkg/storage/replica_state.go#L299)):*
> ```Go
> ) error {
> 	if frozen == storagebase.ReplicaState_FROZEN_UNSPECIFIED {
> 		panic("wat")
> ```

This has been removed.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9883#-:-KTy-0d2LoY5q7mk6Gku:b-z90wns)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 1 of 3 files at r3, 18 of 37 files at r4, 5 of 5 files at r5.
Review status: all files reviewed at latest revision, 19 unresolved discussions, all commit checks successful.

---

*[pkg/storage/replica_proposal.go, line 74 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/9883#-KTtBoTiJUbur8qJvX64:-KTy9hZ9gcF_-q5Ga3yR:bqpzcdw) ([raw file](https://github.com/cockroachdb/cockroach/blob/71bd7c09685cffe1429d6b23990b76ed569c5498/pkg/storage/replica_proposal.go#L74)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> I explicitly avoided `executed` because that  will give you the wrong impression. How about "evaluated"?
</details>
"evaluated"++

---

*[pkg/storage/replica_proposal.go, line 105 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/9883#-KTtCNhAnQztCAXYhKhW:-KTy9pzJolHp1nXyKlez:b6gzewy) ([raw file](https://github.com/cockroachdb/cockroach/blob/71bd7c09685cffe1429d6b23990b76ed569c5498/pkg/storage/replica_proposal.go#L105)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> I want to make absolutely clear that the RHS is not supposed to be used any more, which I don't think the others do quite as clearly. What do you not like about `Absorb`?
</details>
"Absorb" is just a rarely used word, and it doesn't convey to me that the RHS should no longer be used. I like "Merge" too; it's not clear to me that it would be so tempting to reuse the RHS after merging it. "MergeAndDestroy"?

---

*[pkg/storage/replica_proposal.go, line 73 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/9883#-KTy9cTaWbiedr94gpHy:-KTy9cTaWbiedr94gpHz:bj536y8) ([raw file](https://github.com/cockroachdb/cockroach/blob/97e97b4a6c79c93041fb1335f6546f681a5bd994/pkg/storage/replica_proposal.go#L73)):*
> ```Go
> 	// Call maybeAddToReplicaGCQueue.
> 	addToReplicaGCQueue bool
> 	// Call TODO
> ```

There's still a TODO here.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9883#-:-KTyETGcgq2VApvkiu4N:bnfp4nl)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Review status: 16 of 19 files reviewed at latest revision, 19 unresolved discussions, some commit checks pending.

---

*[pkg/storage/replica_proposal.go, line 74 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/9883#-KTtBoTiJUbur8qJvX64:-KTyH_Xy6IV9DNjsejbG:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/71bd7c09685cffe1429d6b23990b76ed569c5498/pkg/storage/replica_proposal.go#L74)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> "evaluated"++
</details>
Done.

---

*[pkg/storage/replica_proposal.go, line 105 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/9883#-KTtCNhAnQztCAXYhKhW:-KTyFmqkQUadPFZWjkey:b-f0h0r8) ([raw file](https://github.com/cockroachdb/cockroach/blob/71bd7c09685cffe1429d6b23990b76ed569c5498/pkg/storage/replica_proposal.go#L105)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> "Absorb" is just a rarely used word, and it doesn't convey to me that the RHS should no longer be used. I like "Merge" too; it's not clear to me that it would be so tempting to reuse the RHS after merging it. "MergeAndDestroy"?
</details>
I was about to give in and rename it to `Merge`, but there's already a `Merge` field on the struct. Settled for `MergeAndDestroy`.

---

*[pkg/storage/replica_proposal.go, line 73 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/9883#-KTy9cTaWbiedr94gpHy:-KTyFbMgR9yJl7p2lpV5:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/97e97b4a6c79c93041fb1335f6546f681a5bd994/pkg/storage/replica_proposal.go#L73)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> There's still a TODO here.
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9883)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: 16 of 19 files reviewed at latest revision, 12 unresolved discussions, all commit checks successful.

---

*[pkg/storage/replica_proposal.go, line 105 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/9883#-KTtCNhAnQztCAXYhKhW:-KTyKbWd4Vb_xkSJcP0z:b-6cj4jw) ([raw file](https://github.com/cockroachdb/cockroach/blob/71bd7c09685cffe1429d6b23990b76ed569c5498/pkg/storage/replica_proposal.go#L105)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> I was about to give in and rename it to `Merge`, but there's already a `Merge` field on the struct. Settled for `MergeAndDestroy`.
</details>
`Assimilate`? I don't really like that either, but need to throw out the Star Trek reference.

There is a `Proposal.Merge` that is inherited from `ReplicatedProposalData.Merge`. I think it is fine, perhaps even desirable, to override that one.

---

*[pkg/storage/replica_proposal.go, line 189 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/9883#-KTtClqxX-0QVP52pH09:-KTyL0puhdcnzxKEGaU3:b-ytq8sy) ([raw file](https://github.com/cockroachdb/cockroach/blob/71bd7c09685cffe1429d6b23990b76ed569c5498/pkg/storage/replica_proposal.go#L189)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> Actually I lied. This is what it should look like:
> 
> ```
> BenchmarkClosureAlloc-4        	20000000       	       103 ns/op       	       1 B/op  	       1 allocs/op
> BenchmarkClosureFunction-4     	2000000000     	         1.07 ns/op    	       0 B/op  	       0 allocs/op
> BenchmarkClosureInline-4       	2000000000     	         0.30 ns/op    	       0 B/op  	       0 allocs/op
> PASS
> ok     	github.com/tschottdorf/goplay/closure_alloc    	5.095s
> ```
> 
> My previous code declared `lhs` outside of the `b.N` loop, hiding the fact that it goes on the heap.
> 
> Moved this out to a static function.
</details>
Oh, very interesting.

---

*[storage/replica_command.go, line 1779 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/9883#-KTtx9gnbLwTMMkjWcf0:-KTyLBuBIedWF7X7C781:bnb7i41) ([raw file](https://github.com/cockroachdb/cockroach/blob/2ac09ca5612f018f081db4f4f2e4ed238f19d92c/storage/replica_command.go#L1779)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> `maybeGossipNodeLiveness` takes a span argument (roughly, it only gossips if it's the leader, but will also only gossip for the key range which either belongs to a single node's record or all records).
</details>
Yeah, but it looks like the gossip key range is static, similar to the system config key range. Feel free to ignore or punt to a future PR.

---

*[storage/replica_command.go, line 2778 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/9883#-KTtxTHzY-UcijsVNbt_:-KTyLJsE-P8j09qOlWuR:bxh7wwp) ([raw file](https://github.com/cockroachdb/cockroach/blob/2ac09ca5612f018f081db4f4f2e4ed238f19d92c/storage/replica_command.go#L2778)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> It would have to be
> 
> ```go
> 	pd := ProposalData{
> 	  ReplicatedProposalData: ReplicatedProposalData{
> 		BlockReads: true,
> 	  }
> 	}
> 
> ```
> 
> which is really noisy.
</details>
Agreed. Didn't realize that was embedded in `ReplicatedProposalData`. Carry on.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9883)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Review status: 16 of 19 files reviewed at latest revision, 12 unresolved discussions, all commit checks successful.

---

*[pkg/storage/replica_proposal.go, line 105 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/9883#-KTtCNhAnQztCAXYhKhW:-KTyNiZ_MMoKB98t1YA9:b-m4qf1k) ([raw file](https://github.com/cockroachdb/cockroach/blob/71bd7c09685cffe1429d6b23990b76ed569c5498/pkg/storage/replica_proposal.go#L105)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> `Assimilate`? I don't really like that either, but need to throw out the Star Trek reference.
> 
> There is a `Proposal.Merge` that is inherited from `ReplicatedProposalData.Merge`. I think it is fine, perhaps even desirable, to override that one.
</details>
You can't override it.

---

*[pkg/storage/replica_proposal.go, line 189 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/9883#-KTtClqxX-0QVP52pH09:-KTyNqJDFjL1q200L2BE:btrwxiz) ([raw file](https://github.com/cockroachdb/cockroach/blob/71bd7c09685cffe1429d6b23990b76ed569c5498/pkg/storage/replica_proposal.go#L189)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> Oh, very interesting.
</details>
I think we knew that before: https://github.com/tschottdorf/goplay/blob/master/alloc_tests/alloc_test.go#L29-L37

---

*[storage/replica_command.go, line 1779 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/9883#-KTtx9gnbLwTMMkjWcf0:-KTyNMkM1c0TvecpwJEx:b-ea9mxf) ([raw file](https://github.com/cockroachdb/cockroach/blob/2ac09ca5612f018f081db4f4f2e4ed238f19d92c/storage/replica_command.go#L1779)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> Yeah, but it looks like the gossip key range is static, similar to the system config key range. Feel free to ignore or punt to a future PR.
</details>
It is here because this is the "all records" case, there's also this code elsewhere (in the commit triggers):

```go
			if nlSpan := ct.ModifiedSpanTrigger.NodeLivenessSpan; nlSpan != nil {
				trigger = updateTrigger(trigger, &PostCommitTrigger{
					maybeGossipNodeLiveness: nlSpan,
				})
			}

```

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9883)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 12 of 23 files at r1, 1 of 3 files at r3, 17 of 37 files at r4, 3 of 5 files at r5, 3 of 3 files at r6.
Review status: all files reviewed at latest revision, 24 unresolved discussions, all commit checks successful.

---

*[pkg/storage/client_raft_test.go, line 1466 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/9883#-KTyL8AdRJkQv4SsaPu3:-KTyL8AdRJkQv4SsaPu4:b-lzz3gp) ([raw file](https://github.com/cockroachdb/cockroach/blob/555bc8c011beabf95d086b47bff364ebb74c13d5/pkg/storage/client_raft_test.go#L1466)):*
> ```Go
> 	rangeID2 := store0.LookupReplica(roachpb.RKey(key), nil).RangeID
> 	if rangeID2 == rangeID {
> 		t.Fatalf("got same range id after split")
> ```

optional: `Fatal` while you're here

---

*[pkg/storage/replica.go, line 2850 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/9883#-KTxXXaxzkM26QJeThCA:-KTyTWspClFAZWnWVel6:bco3a83) ([raw file](https://github.com/cockroachdb/cockroach/blob/2da5711d902fa5ad6179a198c80368074934bc5d/pkg/storage/replica.go#L2850)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> I realized that this is actually best kept in LocalProposalData. Moved it there.
</details>
What moved?

---

*[pkg/storage/replica.go, line 2933 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/9883#-KTxX_f35HRwiP-9A5bK:-KTyTlJZU3jAx6kA0C1x:b-rvsv5r) ([raw file](https://github.com/cockroachdb/cockroach/blob/2da5711d902fa5ad6179a198c80368074934bc5d/pkg/storage/replica.go#L2933)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> Added this assertion.
</details>
expected non empty but then you discard it? why?

---

*[pkg/storage/replica.go, line 2843 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/9883#-KTyTSiloFegfi9HTqtV:-KTyTSimzAHFkihcwtQm:b-9x8ndf) ([raw file](https://github.com/cockroachdb/cockroach/blob/555bc8c011beabf95d086b47bff364ebb74c13d5/pkg/storage/replica.go#L2843)):*
> ```Go
> 	//
> 	// TODO(tschottdorf): this assertion should never fire (as most assertions)
> 	// and it should be removed after 2016/12/01.
> ```

iso 8601 brah. 2016-12-01

---

*[pkg/storage/replica.go, line 2887 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/9883#-KTyT_pwjOzW1v9upirI:-KTyT_pwjOzW1v9upirJ:bqyosqn) ([raw file](https://github.com/cockroachdb/cockroach/blob/555bc8c011beabf95d086b47bff364ebb74c13d5/pkg/storage/replica.go#L2887)):*
> ```Go
> 		// can be read from the WriteBatch, but there may be reasons to pass
> 		// it with propResult. We'll see.
> 		if (pd.State == storagebase.ReplicaState{}) {
> ```

when is this condition true?

---

*[pkg/storage/replica_command.go, line 840 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/9883#-KTyRILs8YnjyT053dFS:-KTyRILs8YnjyT053dFT:blvcphv) ([raw file](https://github.com/cockroachdb/cockroach/blob/555bc8c011beabf95d086b47bff364ebb74c13d5/pkg/storage/replica_command.go#L840)):*
> ```Go
> 	}
> 
> 	if ct.GetSplitTrigger() != nil {
> ```

this is pretty awkward (even before your change). we should make these triggers use `onlyone`.

---

*[pkg/storage/replica_command.go, line 1611 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/9883#-KTyRTq36M6W5LoPMyTr:-KTyRTq36M6W5LoPMyTs:b-u1ga4i) ([raw file](https://github.com/cockroachdb/cockroach/blob/555bc8c011beabf95d086b47bff364ebb74c13d5/pkg/storage/replica_command.go#L1611)):*
> ```Go
> 
> func newFailedLeaseTrigger() ProposalData {
> 	var trigger = ProposalData{}
> ```

you found the least used declaration form!

---

*[pkg/storage/replica_command.go, line 1778 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/9883#-KTyRaQn4fEpR5v056AH:-KTyRaQog9Ce3P_q257K:b-l9ospt) ([raw file](https://github.com/cockroachdb/cockroach/blob/555bc8c011beabf95d086b47bff364ebb74c13d5/pkg/storage/replica_command.go#L1778)):*
> ```Go
> 	pd.State.Lease = &lease
> 	t := true
> 	pd.leaseMetricsResult = &t
> ```

i believe you can use `proto.Bool(true)`.

---

*[pkg/storage/replica_proposal.go, line 537 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9883#-KTogjyrENJQvunbTC3G-r1-537:-KTyU8Wa_8O1abFKEG3V:b-42ylyv) ([raw file](https://github.com/cockroachdb/cockroach/blob/028b1d0071ca6aa5613dc831797b68b2aebd1066/pkg/storage/replica_proposal.go#L537)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> The structs couldn't be compared directly because of the nested `[]intentsWithArg`, but I've made that `*[]intentsWithArg` which is mildly awkward but allows the comparison, which I think is worth the boilerplate it avoids.
</details>
Bummer.

---

*[pkg/storage/replica_proposal.go, line 56 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/9883#-KTyUKmH9Bpg3Tsv1gEi:-KTyUKmH9Bpg3Tsv1gEj:b-ms6uw6) ([raw file](https://github.com/cockroachdb/cockroach/blob/555bc8c011beabf95d086b47bff364ebb74c13d5/pkg/storage/replica_proposal.go#L56)):*
> ```Go
> 	// its non-local intents up the stack.
> 	//
> 	// This is a pointer to allow the zero values to be compared.
> ```

not just zero values =/

---

*[pkg/storage/replica_state.go, line 299 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/9883#-KTxYIIp4JM4saEfSZoV:-KTyQ-LjdOt7jfuRmWU-:b-nzr0l3) ([raw file](https://github.com/cockroachdb/cockroach/blob/2da5711d902fa5ad6179a198c80368074934bc5d/pkg/storage/replica_state.go#L299)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> This has been removed.
</details>
has it?

---

*[pkg/storage/store_test.go, line 2020 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/9883#-KTyQpUu6v90Y7PCkYrg:-KTyQpUu6v90Y7PCkYrh:baza683) ([raw file](https://github.com/cockroachdb/cockroach/blob/555bc8c011beabf95d086b47bff364ebb74c13d5/pkg/storage/store_test.go#L2020)):*
> ```Go
> 	store := tc.store
> 
> 	assertFrozen := func(b bool) {
> ```

this is pretty hard to follow now; can you make this the enum type?

---

*[pkg/storage/storagebase/proposer_kv.proto, line 63 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/9883#-KTyPi8RGkru4DB1dgd1:-KTyPi8SggLrEFHpOm_8:b-yhf92h) ([raw file](https://github.com/cockroachdb/cockroach/blob/555bc8c011beabf95d086b47bff364ebb74c13d5/pkg/storage/storagebase/proposer_kv.proto#L63)):*
> ```protobuf
>   // which additional side effects (for instance on a descriptor update).
>   optional storage.storagebase.ReplicaState state = 2 [(gogoproto.nullable) = false];
>   optional Split split = 3;
> ```

should these two guys be `onlyone`?

---

*[pkg/storage/storagebase/proposer_kv.proto, line 65 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/9883#-KTyPkqOroxFGlKo59Sg:-KTyPkqOroxFGlKo59Sh:bq8afnp) ([raw file](https://github.com/cockroachdb/cockroach/blob/555bc8c011beabf95d086b47bff364ebb74c13d5/pkg/storage/storagebase/proposer_kv.proto#L65)):*
> ```protobuf
>   optional Split split = 3;
>   optional Merge merge = 4;
>   optional roachpb.ComputeChecksumRequest compute_checksum = 5;
> ```

comment?

---

*[pkg/storage/storagebase/state.proto, line 78 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/9883#-KTyS5kH_gSVLc1L-f4p:-KTyS5kH_gSVLc1L-f4q:b-pztzv5) ([raw file](https://github.com/cockroachdb/cockroach/blob/555bc8c011beabf95d086b47bff364ebb74c13d5/pkg/storage/storagebase/state.proto#L78)):*
> ```protobuf
>       (gogoproto.customname) = "TxnSpanGCThreshold"];
>   // Use an enum because proto3 does not give nullable primitive values, and we
>   // want to need to be able to send ProposalData which don't specify frozen.
> ```

"want to need to"

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9883)*
<!-- Sent from Reviewable.io -->

tbg committed:
stability: prepare protos for proposer-evaluted KV

As a page-in task for proposer-evaluted KV, this establishes structures
to be used for sending the new types of proposals over the wire. In short,
they consist of three components:

- Local state only available to the leader (intents to resolve, ...)
- A replicated state update for all Replicas (split/merge triggers,
  range descriptor updates, log queue truncation parameters, increased
  GC thresholds, ...). This always contains the new applied index and
  stats, so the structure is intentionally not optional (but most of
  its contents are, so it is not a large struct and serializes well enough).
- A RocksDB WriteBatch (i.e. the actual data to be written to disk).

There are some TODOs I would like to address before merging (moving code,
adding tests and assertions), but this should make for a good high-level
review as is.

There are a lot of changes; I recommend focusing on
`storage/replica_proposal.go` as well as `storage/storagebase`.

tschottdorf commented:
TFTRs!

---

Review status: all files reviewed at latest revision, 23 unresolved discussions, all commit checks successful.

---

*[pkg/storage/client_raft_test.go, line 1466 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/9883#-KTyL8AdRJkQv4SsaPu3:-KTyZShv8z8NTBfAIA5X:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/555bc8c011beabf95d086b47bff364ebb74c13d5/pkg/storage/client_raft_test.go#L1466)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> optional: `Fatal` while you're here
</details>
Done.

---

*[pkg/storage/replica.go, line 2850 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/9883#-KTxXXaxzkM26QJeThCA:-KTyZdSmv69V4IWK7NPa:b-hncngf) ([raw file](https://github.com/cockroachdb/cockroach/blob/2da5711d902fa5ad6179a198c80368074934bc5d/pkg/storage/replica.go#L2850)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> What moved?
</details>
`delta enginepb.MVCCStats`

---

*[pkg/storage/replica.go, line 2933 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/9883#-KTxX_f35HRwiP-9A5bK:-KTy_ljr8AMLlUHenmFf:bevkp47) ([raw file](https://github.com/cockroachdb/cockroach/blob/2da5711d902fa5ad6179a198c80368074934bc5d/pkg/storage/replica.go#L2933)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> expected non empty but then you discard it? why?
</details>
It's expected empty at this point. Just asserting that nobody sets this earlier.

---

*[pkg/storage/replica.go, line 2843 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/9883#-KTyTSiloFegfi9HTqtV:-KTyZcZyqRvJt9Exejub:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/555bc8c011beabf95d086b47bff364ebb74c13d5/pkg/storage/replica.go#L2843)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> iso 8601 brah. 2016-12-01
</details>
Done.

---

*[pkg/storage/replica.go, line 2887 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/9883#-KTyT_pwjOzW1v9upirI:-KTyZtzBNtQEtabk9MMz:b7sibdr) ([raw file](https://github.com/cockroachdb/cockroach/blob/555bc8c011beabf95d086b47bff364ebb74c13d5/pkg/storage/replica.go#L2887)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> when is this condition true?
</details>
Never! Thanks for noticing. This saves a lot of work.

---

*[pkg/storage/replica_command.go, line 1611 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/9883#-KTyRTq36M6W5LoPMyTr:-KTya0uoyL4LtNbJsoyM:bcfzt63) ([raw file](https://github.com/cockroachdb/cockroach/blob/555bc8c011beabf95d086b47bff364ebb74c13d5/pkg/storage/replica_command.go#L1611)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> you found the least used declaration form!
</details>
Haha. Amazing no linter catches it. Changed.

---

*[pkg/storage/replica_command.go, line 1778 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/9883#-KTyRaQn4fEpR5v056AH:-KTyaAIQRMnCcyxx64hB:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/555bc8c011beabf95d086b47bff364ebb74c13d5/pkg/storage/replica_command.go#L1778)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> i believe you can use `proto.Bool(true)`.
</details>
Done.

---

*[pkg/storage/replica_proposal.go, line 56 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/9883#-KTyUKmH9Bpg3Tsv1gEi:-KTyamq7tCscdMhbbe2V:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/555bc8c011beabf95d086b47bff364ebb74c13d5/pkg/storage/replica_proposal.go#L56)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> not just zero values =/
</details>
Done.

---

*[pkg/storage/replica_state.go, line 299 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/9883#-KTxYIIp4JM4saEfSZoV:-KTyasZh5CGkOacFr8hS:btvivs9) ([raw file](https://github.com/cockroachdb/cockroach/blob/2da5711d902fa5ad6179a198c80368074934bc5d/pkg/storage/replica_state.go#L299)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> has it?
</details>
Well, not in this PR until now. Done.

---

*[pkg/storage/store_test.go, line 2020 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/9883#-KTyQpUu6v90Y7PCkYrg:-KTyfrNpMJU1g89qqULQ:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/555bc8c011beabf95d086b47bff364ebb74c13d5/pkg/storage/store_test.go#L2020)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> this is pretty hard to follow now; can you make this the enum type?
</details>
Done.

---

*[pkg/storage/storagebase/proposer_kv.proto, line 63 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/9883#-KTyPi8RGkru4DB1dgd1:-KTyfxxR8oJsKV8IvPLU:b-5vxtmn) ([raw file](https://github.com/cockroachdb/cockroach/blob/555bc8c011beabf95d086b47bff364ebb74c13d5/pkg/storage/storagebase/proposer_kv.proto#L63)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> should these two guys be `onlyone`?
</details>
Maybe, but I'll punt now.

---

*[pkg/storage/storagebase/proposer_kv.proto, line 65 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/9883#-KTyPkqOroxFGlKo59Sg:-KTygAZSFulvHAuxX7tb:bu3fl9s) ([raw file](https://github.com/cockroachdb/cockroach/blob/555bc8c011beabf95d086b47bff364ebb74c13d5/pkg/storage/storagebase/proposer_kv.proto#L65)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> comment?
</details>
I'll add a comment when I've trimmed this (TODO added).

---

*[pkg/storage/storagebase/state.proto, line 78 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/9883#-KTyS5kH_gSVLc1L-f4p:-KTygIgOR2iBZL9OqDDg:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/555bc8c011beabf95d086b47bff364ebb74c13d5/pkg/storage/storagebase/state.proto#L78)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> "want to need to"
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9883#-:-KTyiuQ9yPvCl9S6p35G:bz5mair)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Review status: 9 of 19 files reviewed at latest revision, 23 unresolved discussions, some commit checks pending.

---

*[pkg/storage/replica_command.go, line 840 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/9883#-KTyRILs8YnjyT053dFS:-KTyj3TzSGy6uInUUVAW:b-58swmn) ([raw file](https://github.com/cockroachdb/cockroach/blob/555bc8c011beabf95d086b47bff364ebb74c13d5/pkg/storage/replica_command.go#L840)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> this is pretty awkward (even before your change). we should make these triggers use `onlyone`.
</details>
I'm not opposed, but not inclined to pull on that thread. Could you take a look?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9883)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 1 of 3 files at r3, 3 of 37 files at r4, 10 of 10 files at r7.
Review status: all files reviewed at latest revision, 10 unresolved discussions, some commit checks pending.

---

*[pkg/storage/replica_command.go, line 840 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/9883#-KTyRILs8YnjyT053dFS:-KTykS8duiF9QRlXW3Kb:b-fnpio9) ([raw file](https://github.com/cockroachdb/cockroach/blob/555bc8c011beabf95d086b47bff364ebb74c13d5/pkg/storage/replica_command.go#L840)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> I'm not opposed, but not inclined to pull on that thread. Could you take a look?
</details>
Sure; will do after this goes in.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9883)*
<!-- Sent from Reviewable.io -->

petermattis commented:
FYI, this PR reduced performance for a 1m `block_writer` run against a local 1-node cluster by 37%. Any guesses as to why?

It's unfortunate that the only way we notice these perf regressions is manually. We really need to fix that.

petermattis commented:
The perf difference is entirely do the calls to `assertState`. We call `assertState` in `Replica.processRaftCommand` for every write operation and `Replica.handleProposalData` also calls `assertState` on almost every operation. 

@tschottdorf Was this change in behavior intentional? A 37% performance it is pretty significant.

tschottdorf commented:
No, absolutely not intentional. In fact, I did my best to avoid that call happening all the time, but clearly failed. I'll take a look and send a fix. Thanks for looking at this - we need to get (something like) this on our continuous deployment pipeline asap.

petermattis commented:
Repro steps:

```
~ killall -9 cockroach; rm -fr cockroach-data/ && roachdemo -n 1 -- --logtostderr
...
~ block_writer -duration 1m
```

On my laptop, I see instantaneous performance around ~3000-3500 ops/sec. If I comment out the `assertState` calls the performance grows to ~5000 ops/sec.

tschottdorf commented:
Thanks. Heads up, I'll submit a (trivial) performance sanity check with my fix for this.
